### PR TITLE
Issue2884 pid autotuning

### PR DESCRIPTION
**Pull request by SenHuang19**
This is to address the comments in #2957
@JayHuLBL Can you please review it?

### CODE REVIEW

**Comment by JayHuLBL** [comment]
@SenHuang19 
By running `../bin/verifyFiles.py`, I see following errors:
```
jianjunhu@ubuntu:~/GitFolder/SenHuang19/modelica-buildings/Buildings$ ../bin/verifyFiles.py 
*** Error:  File 'Controls/OBC/Utilities/PIDWithAutotuning/Relay/BaseClasses/package.mo' does not contain required string 'documentation'.
*** Error:  File 'Controls/OBC/Utilities/PIDWithAutotuning/SystemIdentification/FirstOrderTimedelayed/BaseClasses/package.mo' does not contain required string 'documentation'.
*** Error:  File 'Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/BaseClasses/package.mo' does not contain required string 'documentation'.
*** Terminating due to found errors in examined files.
```

File: Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/BaseClasses/Validation/PIDGain.mo [hunk 1]
```diff
@@ -0,0 +1,70 @@
+within Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO.BaseClasses.Validation;
+model PIDGain "Test model for calculating the control gain for a PID controller"
+  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO.BaseClasses.PIDGain
+                                                                             PIDGai
```
Comment (JayHuLBL) [inline Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/BaseClasses/Validation/PIDGain.mo:4]: change to one line.

File: Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/BaseClasses/Validation/PIGain.mo [hunk 2]
```diff
@@ -0,0 +1,70 @@
+within Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO.BaseClasses.Validation;
+model PIGain "Test model for calculating the gain for a PI controller"
+  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO.BaseClasses.PIGain
+                                                                            PIGai
```
Comment (JayHuLBL) [inline Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/BaseClasses/Validation/PIGain.mo:4]: one line.

File: Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/BaseClasses/package.mo [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+          fillColor={248,248,248},
+          fillPattern=FillPattern.HorizontalCylinder,
+          extent={{-100,-100},{100,100}},
+          radius=25.0),
+        Ellipse(
+          extent={{-30,-32},{30,28}},
+          lineColor={128,128,128},
+          fillColor={255,255,255},
+          fillPattern=FillPattern.Solid)}));
+end BaseClasses;
```
Comment (JayHuLBL) [inline Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/BaseClasses/package.mo:16]: missing `Documentation` section.

File: Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo [hunk 4]
```diff
@@ -30,11 +30,12 @@ block PI "Identify control gain and integral time of the PI controller"
     "Time constant signal for the integral term"
     annotation (Placement(transformation(extent={{100,-50},{140,-10}}),
         iconTransformation(extent={{100,-80},{140,-40}})));
-  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO.PIGain gai
-    "Calculate the control gain"
+protected
+  BaseClasses.PIGain                                                        gai
```
Comment (JayHuLBL) [inline Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo:34]: use full path, remove the empty spaces.

File: Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo [hunk 5]
```diff
@@ -30,11 +30,12 @@ block PI "Identify control gain and integral time of the PI controller"
...
+    "Block that calculates the control gain"
     annotation (Placement(transformation(extent={{-10,20},{10,40}})));
-  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO.PIIntegralTime
-    intTim "Calculate the integral time"
+  BaseClasses.PIIntegralTime
```
Comment (JayHuLBL) [inline Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo:37]: full path, one line.

File: Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo [hunk 6]
```diff
@@ -62,7 +63,40 @@ equation
         Text(
           extent={{-100,140},{100,100}},
           textString="%name",
-          textColor={0,0,255})}), Diagram(coordinateSystem(preserveAspectRatio=false)),
+          textColor={0,0,255}),
+        Text(
+          visible=(controllerType == Buildings.Controls.OBC.CDL.Types.SimpleController.PI),
```
Comment (JayHuLBL) [inline Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo:68]: I see this line in various models. It seems should not be here (and the model does not have parameter `controllerType`).
Please check other models as well.

File: Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PID.mo [hunk 7]
```diff
@@ -27,23 +27,24 @@ block PID "Identify control gain, integral time, and derivative time of the PID
...
     final unit="s",
     min=100*Buildings.Controls.OBC.CDL.Constants.eps)
-    "Connector for time constant signal for the derivative term"
+    "Time constant signal for the derivative term"
     annotation (Placement(transformation(extent={{100,-80},{140,-40}}),
         iconTransformation(extent={{100,-90},{140,-50}})));
-  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.AutoTuner.AMIGO.PIDGain gai
-    "Calculate the control gain"
+protected
+  BaseClasses.PIDGain                                                        gai
```
Comment (JayHuLBL) [inline Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PID.mo:40]: full path, one line.

File: Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/ResponseProcess.mo [hunk 8]
```diff
@@ -37,19 +37,20 @@ block ResponseProcess
...
         iconTransformation(extent={{100,-20},{140,20}})));
-  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Relay.OnOffPeriod onOffPer
-    "Calculate the length of the on period and the off period"
+protected
+  BaseClasses.OnOffPeriod                                              onOffPer
+    "Block that calculates the length of the on period and the off period"
     annotation (Placement(transformation(extent={{-80,0},{-60,20}})));
-  Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Relay.HalfPeriodRatio halPerRatio
-    "Calculate the half period ratio"
+  BaseClasses.HalfPeriodRatio                                              halPerRatio
```
Comment (JayHuLBL) [inline Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/ResponseProcess.mo:44]: full path, one line.

**Review by JayHuLBL — CHANGES_REQUESTED** [review]
@SenHuang19 Please see inline comments.

**Comment by SenHuang19** [comment]
> @SenHuang19
> 
> * By running `../bin/verifyFiles.py`, I see following errors:
> 
> ```
> jianjunhu@ubuntu:~/GitFolder/SenHuang19/modelica-buildings/Buildings$ ../bin/verifyFiles.py 
> *** Error:  File 'Controls/OBC/Utilities/PIDWithAutotuning/Relay/BaseClasses/package.mo' does not contain required string 'documentation'.
> *** Error:  File 'Controls/OBC/Utilities/PIDWithAutotuning/SystemIdentification/FirstOrderTimedelayed/BaseClasses/package.mo' does not contain required string 'documentation'.
> *** Error:  File 'Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/BaseClasses/package.mo' does not contain required string 'documentation'.
> *** Terminating due to found errors in examined files.
> ```
> 
> * Also, following hyperlinks are broken:
> 
> ```
> Error: Can not resolve link "modelica://Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Relay.HalfPeriodRatio" in Buildings.Controls.OBC.Utilities.PIDWithAutotuning.FirstOrderAMIGO
> Error: Can not resolve link "modelica://Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Relay.Validation.NormalizedTimeDelay" in Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Relay.BaseClasses.NormalizedTimeDelay
> Error: Can not resolve link "modelica://Buildings.Controls.OBC.Utilities.PIDWithAutotuning.SystemIdentification.FirstOrderTimedelayed.Validation.TimeConstantDelay" in Buildings.Controls.OBC.Utilities.PIDWithAutotuning.SystemIdentification.FirstOrderTimedelayed.BaseClasses.TimeConstantDelay
> Error: Can not resolve link "modelica://Buildings.Controls.OBC.Utilities.PIDWithAutotuning.SystemIdentification.FirstOrderTimedelayed.BaseClassess.Gain" in Buildings.Controls.OBC.Utilities.PIDWithAutotuning.SystemIdentification.FirstOrderTimedelayed.BaseClasses.Validation.Gain
> Error: Can not resolve link "modelica://Buildings.Controls.OBC.Utilities.PIDWithAutotuning.SystemIdentification.FirstOrderTimedelayed.BaseClassess.TimeConstantDelay" in Buildings.Controls.OBC.Utilities.PIDWithAutotuning.SystemIdentification.FirstOrderTimedelayed.BaseClasses.Validation.TimeConstantDelay
> Generated documentation for Buildings.Controls.OBC.Utilities.PIDWithAutotuning in /home/jianjunhu/GitFolder/SenHuang19/modelica-buildings/Buildings/help/Buildings_Controls_OBC_Utilities_PIDWithAutotuning.html
> ```
> 
> * in `Buildings.Controls.OBC.Utilities.PIDWithAutotuning.FirstOrderAMIGO`, for the Icon layer text string "PI", change the `textString="PI"` to `textString= if controllerType == Buildings.Controls.OBC.CDL.Types.SimpleController.PID then "PID" else "PI"` and remove the `visible=(controllerType == Buildings.Controls.OBC.CDL.Types.SimpleController.PI),`.
> * please also see the inline comments below.

- The document is added
- The broken links are fixed
- The icon is updated

File: Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/BaseClasses/Validation/PIDGain.mo (hunk 1 above)
Comment (SenHuang19) [reply Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/BaseClasses/Validation/PIDGain.mo]: Fix the format.

File: Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/BaseClasses/package.mo (hunk 3 above)
Comment (SenHuang19) [reply Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/BaseClasses/package.mo:22]: Add <code>Documentation</code>.

File: Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo (hunk 5 above)
Comment (SenHuang19) [reply Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo]: The full path is used and the format is fixed

File: Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo (hunk 6 above)
Comment (SenHuang19) [reply Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PI.mo]: Remove this line from this model and other models.

File: Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PID.mo (hunk 7 above)
Comment (SenHuang19) [reply Buildings/Controls/OBC/Utilities/PIDWithAutotuning/AutoTuner/AMIGO/PID.mo]: The full path is used and the format is fixed.

File: Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/ResponseProcess.mo (hunk 8 above)
Comment (SenHuang19) [reply Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/ResponseProcess.mo]: The full path is used and the format is fixed

**Comment by SenHuang19** [comment]
> @SenHuang19 Please see inline comments.
@JayHuLBL thanks a lot for the comments. 
Most of them are due to the recent restructure of the sub-packages.
I believe I addressed all of your comments.
Can you please review it again?

**Review by JayHuLBL — APPROVED** [review]

**Pull request merged**